state/runtime/allowlist/allowlist.go

```diff
@@ -0,0 +1,178 @@
+
+	res := &runtime.ExecutionResult{
+		ReturnValue: ret,
+		GasUsed:     gasUsed,
+		GasLeft:     c.Gas - gasUsed,
+		Err:         err,
+	}
+
+	return res
+}
+
+var (
+	errNoFunctionSignature = fmt.Errorf("input is too short for a function call")
+	errInputTooShort       = fmt.Errorf("wrong input size, expected 32")
+	errFunctionNotFound    = fmt.Errorf("function not found")
+	errWriteProtection     = fmt.Errorf("write protection")
+	errNotAuth             = fmt.Errorf("not auth")
+)
+
+func (a *AllowList) runInputCall(caller types.Address, input []byte,
+	gas uint64, isStatic bool) ([]byte, uint64, error) {
+	// decode the function signature from the input
+	if len(input) < 4 {
+		return nil, 0, errNoFunctionSignature
+	}
+
+	sig, inputBytes := input[:4], input[4:]
+
+	// all the functions have the same input (i.e. tuple(address)) which
+	// in abi gets codified as a 32 bytes array with the first 20 bytes
+	// encoding the address
+	if len(inputBytes) != 32 {
+		return nil, 0, errInputTooShort
+	}
+
+	var gasUsed uint64
+
+	consumeGas := func(gasConsume uint64) error {
+		if gas < gasConsume {
+			return runtime.ErrOutOfGas
+		}
+
+		gasUsed = gasConsume
+
+		return nil
+	}
+
+	inputAddr := types.BytesToAddress(inputBytes)
+
+	if bytes.Equal(sig, ReadAllowListFunc.ID()) {
+		if err := consumeGas(readAllowListCost); err != nil {
+			return nil, 0, err
+		}
+
+		// read operation
+		role := a.GetRole(inputAddr)
+
+		return role.Bytes(), gasUsed, nil
+	}
+
+	// write operation
+	var updateRole Role
+	if bytes.Equal(sig, SetAdminFunc.ID()) {
+		updateRole = AdminRole
+	} else if bytes.Equal(sig, SetEnabledSignatureFunc.ID()) {
+		updateRole = EnabledRole
+	} else if bytes.Equal(sig, SetNoneFunc.ID()) {
+		updateRole = NoRole
+	} else {
+		return nil, 0, errFunctionNotFound
+	}
+
+	if err := consumeGas(writeAllowListCost); err != nil {
+		return nil, gasUsed, err
+	}
+
+	// we cannot perform any write operation if the call is static
+	if isStatic {
+		return nil, gasUsed, errWriteProtection
+	}
+
+	// Only Admin accounts can modify the role of other accounts
+	addrRole := a.GetRole(caller)
+	if addrRole != AdminRole {
+		return nil, gasUsed, errNotAuth
+	}
+
+	a.SetRole(inputAddr, updateRole)
+
+	return nil, gasUsed, nil
+}
+
+func (a *AllowList) SetRole(addr types.Address, role Role) {
+	a.state.SetState(a.addr, types.BytesToHash(addr.Bytes()), types.Hash(role))
+}
+
+func (a *AllowList) GetRole(addr types.Address) Role {
+	res := a.state.GetStorage(a.addr, types.BytesToHash(addr.Bytes()))
+
+	return Role(res)
+}
+
+type Role types.Hash
+
+var (
+	NoRole      Role = Role(types.StringToHash("0x0000000000000000000000000000000000000000000000000000000000000000"))
+	EnabledRole Role = Role(types.StringToHash("0x0000000000000000000000000000000000000000000000000000000000000001"))
+	AdminRole   Role = Role(types.StringToHash("0x0000000000000000000000000000000000000000000000000000000000000002"))
```

Comment:
Just use a shorthand, it should be enough. WDYT?

```suggestion
	NoRole      Role = Role(types.StringToHash("0x0"))
	EnabledRole Role = Role(types.StringToHash("0x1"))
	AdminRole   Role = Role(types.StringToHash("0x2"))
```

Reply:
Well, this is a hack that `StringToHash` does. I prefer to be explicit here so that we are fully aware that `0x1` is `0x0...1` and not `0x01...0`.